### ISSUE TITLE
gradient_clipping.mojo uses scalar loops for norm computation — should use SIMD

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#4911]
## Problem

`shared/training/gradient_clipping.mojo` computes gradient norms and applies clipping using scalar element-by-element loops (`_get_float64` / `_set_float64` per element). This is called every training step, making it a performance bottleneck for large models.

### Affected functions

- `compute_gradient_norm_list()` — iterates every element of every gradient tensor with `_get_float64(j)`
- `clip_gradients_by_global_norm()` — scales every element with `_get_float64(j)` and `_set_float64(j, ...)`
- `clip_gradients_per_param()` — same pattern per parameter

### Performance impact

For a model with 10M parameters, each training step performs 10M+ scalar memory accesses for norm computation and another 10M+ for clipping. SIMD vectorization would reduce this by 4-16x depending on dtype and hardware.

### Recommended fix

1. Use typed SIMD pointers (`bitcast[Float32]()`) and `vectorize[]` for norm accumulation.
2. Use SIMD multiply-and-store for the clipping scale operation.
3. Avoid the `_get_float64` / `_set_float64` indirection which adds unnecessary Float64 conversion overhead for Float32 tensors.
4. Follow the pattern established in `matmul.mojo` and `activation_simd.mojo`.

### Location

- `shared/training/gradient_clipping.mojo`

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

I now have all the context needed. Let me produce the implementation plan.

---

# Implementation Plan: SIMD Vectorization for `gradient_clipping.mojo`

**Issue**: #4911 — `gradient_clipping.mojo` uses scalar loops for norm computation — should use SIMD

## 1. Summary

The `shared/training/gradient_clipping.mojo` module computes gradient norms and applies clipping using scalar `_get_float64(j)` / `_set_float64(j, ...)` element-by-element loops. For a 10M-parameter model, this means 20M+ scalar memory accesses per training step with unnecessary Float64 conversion overhead for Float32 tensors.

**Solution**: Replace scalar loops with SIMD-vectorized inner loops using `algorithm.vectorize`, typed `bitcast` pointers, and `simd_width_of` — following the established patterns in `activation_simd.mojo`, `arithmetic_simd.mojo`, and `matmul.mojo`. This is an **in-place optimization** of 5 functions with no API changes.

**Expected speedup**: 4–8× for Float32, 2–4× for Float64 on the hot-path norm and scale operations.

## 2. Step-by-Step Implementation Tasks

### Task 1: Add SIMD imports

Add the required imports at the top of `gradient_clipping.mojo`:

```mojo
from algorithm import vectorize
from sys.info import simd_width_of
```

**File**: `shared/training/gradient_clipping.mojo` (line 30, after existing imports)

---

### Task 2: Vectorize `compute_gradient_norm_list()`

**Current** (lines 57–68): Scalar loop with `_get_float64(j)`, accumulating `Float64` squared sum.

**New approach**: For each gradient tensor, dispatch on dtype:

- **Float32 path**: `bitcast[Float32]()`, SIMD `load[width]`, compute `vec * vec`, horizontally reduce to `Float64` accumulator.
- **Float64 path**: Same pattern with `bitcast[Float64]()`.
- **Fallback**: Keep scalar `_get_float64` for non-float dtypes.

**Implementation pattern** (follows `arithmetic_simd.mojo` lines 96–112):

```mojo
fn compute_gradient_norm_list(gradients: List[ExTensor]) raises -> Float32:
    var total_norm_sq = Float64(0.0)

    for i in range(len(gradients)):
        var grad = gradients[i]
        if grad._dtype == DType.float32:
            total_norm_sq += _norm_sq_simd_f32(grad)
        elif grad._dtype == DType.float64:
            total_norm_sq += _norm_sq_simd_f64(grad)
        else:
            # Scalar fallback for other dtypes
            for j in range(grad.numel()):
                var val = grad._get_float64(j)
                total_norm_sq += val * val

    return Float32(sqrt(total_norm_sq))
```

**New helper — `_norm_sq_simd_f32`**:

```mojo
@always_inline
fn _norm_sq_simd_f32(tensor: ExTensor) -> Float64:
    comptime simd_width = simd_width_of[DType.float32]()
    var size = tensor._numel
    var ptr = tensor._data.bitcast[Float32]()
    var acc = Float64(0.0)

    @parameter
    fn vectorized_norm[width: Int](idx: Int) unified {mut}:
        var vec = ptr.load[width=width](idx)
        var sq = vec * vec
        # Reduce SIMD vector to scalar and accumulate
        acc += sq.reduce_add().cast[DType.float64]()

    vectorize[simd_width](size, vectorized_norm)
    return acc
```

**New helper — `_norm_sq_simd_f64`**: Same pattern with `Float64` pointers and `simd_width_of[DType.float64]()`.

---

### Task 3: Add SIMD scale helpers

Create `_scale_simd_f32` and `_scale_simd_f64` helpers for in-place scaling (used by both global and per-param clipping):

```mojo
@always_inline
fn _scale_simd_f32(tensor: ExTensor, scale: Float32):
    comptime simd_width = simd_width_of[DType.float32]()
    var size = tensor._numel
    var ptr = tensor._data.bitcast[Float32]()
    var scale_vec = SIMD[DType.float32, 1](scale)  # broadcasts automatically

    @parameter
    fn vectorized_scale[width: Int](idx: Int) unified {mut}:
        var vec = ptr.load[width=width](idx)
        ptr.store[width=width](idx, vec * scale.cast[DType.float32]())

    vectorize[simd_width](size, vectorized_scale)
```

And similarly for `_scale_simd_f64`.

---

### Task 4: Vectorize `clip_gradients_by_global_norm()`

**Current** (lines 113–119): Scalar loop with `_get_float64(j)` and `_set_float64(j, val * clip_coef)`.

**New approach**: After computing `clip_coef`, dispatch each gradient to the SIMD scale helper:

```mojo
if total_norm > max_norm:
    var clip_coef = Float64(max_norm) / (Float64(total_norm) + 1e-6)

    for i in range(len(gradients)):
        var grad = gradients[i]
        if grad._dtype == DType.float32:
            _scale_simd_f32(grad, Float32(clip_coef))
        elif grad._dtype == DType.float64:
            _scale_simd_f64(grad, Float64(clip_coef))
        else:
            for j in range(grad.numel()):
                var val = grad._get_float64(j)
                grad._set_float64(j, val * clip_coef)
```

**Note**: `compute_gradient_norm_list` is already vectorized from Task 2, so both norm computation and scaling benefit.

---

### Task 5: Vectorize `clip_gradients_per_param()`

**Current** (lines 153–171): Two scalar loops per parameter — one for norm, one for scaling.

**New approach**: Reuse `_norm_sq_simd_f32/f64` for norm computation and `_scale_simd_f32/f64` for scaling:

```mojo
for i in range(len(gradients)):
    var grad = gradients[i]
    var param_norm_sq: Float64

    if grad._dtype == DType.float32:
        param_norm_sq = _norm_sq_simd_f32(grad)
    elif grad._dtype == DType.float64:
        param_norm_sq = _norm_sq_simd_f64(grad)
    else:
        # scalar fallback
        ...

    var param_norm = Float64(sqrt(param_norm_sq))
    if param_norm > Float64(max_norm):
        var clip_coef = Float64(max_norm) / (param_norm + 1e-6)
        if grad._dtype == DType.float32:
            _scale_simd_f32(grad, Float32(clip_coef))
        elif grad._dtype == DType.float64:
            _scale_simd_f64(grad, Float64(clip_coef))
        else:
            # scalar fallback
            ...
```

---

### Task 6: Vectorize `clip_gradients_by_value_list()`

**Current** (lines 207–217): Scalar loop with per-element clamping.

**New approach**: Use SIMD `max`/`min` (same pattern as `_relu_simd_float32` in `activation_simd.mojo`):

```mojo
@always_inline
fn _clamp_simd_f32(tensor: ExTensor, min_val: Float32, max_val: Float32):
    comptime simd_width = simd_width_of[DType.float32]()
    var size = tensor._numel
    var ptr = tensor._data.bitcast[Float32]()

    @parameter
    fn vectorized_clamp[width: Int](idx: Int) unified {mut}:
        var vec = ptr.load[width=width](idx)
        # Clamp: max(min_val, min(max_val, vec))
        var clamped = vec.min(SIMD[DType.float32, width](max_val))
                         .max(SIMD[DType.float32, width](min_val))
        ptr.store[width=width](idx, clamped)

    vectorize[simd_width](size, vectorized_clamp)
```

---

### Task 7: Vectorize `compute_gradient_statistics()` (partial)

**Current** (lines 328–367): Scalar loop computing norm, min, max, sum, NaN/Inf counts.

**New approach**: This function has mixed concerns (NaN/Inf detection requires per-element checks). Use a **hybrid** approach:

- SIMD for norm accumulation (`vec * vec` → `reduce_add`)
- SIMD for abs/max/min tracking (using `SIMD.reduce_max()`, `SIMD.reduce_min()`)
- Scalar tail-loop only for NaN/Inf counting (these are rare-path operations)

**Alternative (simpler, recommended for first pass)**: Keep `compute_gradient_statistics()` scalar since it's a monitoring function called infrequently (not every step). Add a `# TODO: SIMD optimization` comment. The hot-path functions (Tasks 2–6) deliver 95%+ of the performance gain.

---

### Task 8: Run existing tests to verify correctness

All 9 existing tests in `test_gradient_clipping.mojo` must pass without modification. The SIMD optimization is a pure internal refactor — the public API and numerical results are unchanged.

---

### Task 9: Add SIMD-specific edge-case tests

Add new test cases to `tests/shared/training/test_gradient_clipping.mojo`:

1. **`test_norm_simd_alignment`**: Tensor sizes that are NOT multiples of SIMD width (e.g., numel=7, 13, 33) to verify `vectorize` handles the tail correctly.
2. **`test_clip_large_tensor`**: Tensor with 10,000+ elements to exercise the SIMD path meaningfully.
3. **`test_clip_float64_tensors`**: Repeat key tests with `DType.float64` tensors.
4. **`test_value_clip_mixed_signs`**: Tensor with mixed positive/negative values for SIMD clamp correctness.

---

## 3. Files to Modify/Create

| File | Action | Description |
|------|--------|-------------|
| `shared/training/gradient_clipping.mojo` | **Modify** | Add SIMD imports; add 6 private helper functions (`_norm_sq_simd_f32`, `_norm_sq_simd_f64`, `_scale_simd_f32`, `_scale_simd_f64`, `_clamp_simd_f32`, `_clamp_simd_f64`); refactor 4 public functions to dispatch to SIMD helpers |
| `tests/shared/training/test_gradient_clipping.mojo` | **Modify** | Add 4+ new test functions for SIMD edge cases; update test count in `main()` |

**No new files needed.** This is a pure in-place optimization.

## 4. Testing Approach

### Correctness Verification (TDD)

1. **Run existing 9 tests first** — they serve as the regression baseline. Zero test changes means the refactor is numerically correct.
2. **Add edge-case tests** for alignment (non-SIMD-width-multiple sizes), large tensors, Float64 path, and mixed-sign value clipping.
3. **Use known-value tensors** with analytically computable norms (e.g., `ones([100])` → norm = 10.0).

### Numerical Accuracy

- Keep `Float64` accumulation for norm computation (avoids Float32 catastrophic cancellation on large tensors). The SIMD inner loop produces `Float32` partial sums that are widened to `Float64` via `cast[DType.float64]()` before accumulation.
- Tolerance thresholds in existing tests (`atol=1e-5`) remain appropriate.

### Build Verification

```bash
just build                    # Verify compilation
just test-group "tests/shared/training" "test_gradient_clipping.mojo"  # Run tests
```

### Performance Validation (manual)

- Create a simple benchmark script with 10M-element tensors
- Compare wall-clock time of norm + clip before/after optimization
- Expected: 4–8× speedup for Float32

## 5. Success Criteria

| Criterion | Metric |
|-----------|--------|
| **All existing tests pass** | 9/9 tests pass with zero modifications |
| **New edge-case tests pass** | 4+ new tests pass (non-aligned sizes, large tensors, float64, mixed signs) |
| **No `_get_float64`/`_set_float64` in hot paths** | The 4 clipping functions use SIMD for float32/float64; scalar fallback only for exotic dtypes |
| **Follows established SIMD patterns** | Uses `bitcast`, `vectorize`, `simd_width_of`, `@always_inline` — matching `arithmetic_simd.mojo` and `activation_simd.mojo` |
| **API unchanged** | All function signatures, return types, and semantics are identical |
| **Numerical accuracy preserved** | Float64 accumulation maintained; results match within existing `atol` thresholds |
| **CI passes** | `just validate` completes successfully |

## Resource Usage

| Resource | Count |
|----------|-------|
| Tool calls used | 9 |
| Tool calls remaining | 41 |
| Steps used | ~5 |
| Steps remaining | ~45 |
| Files read | 6 (gradient_clipping.mojo, activation_simd.mojo, matmul.mojo, arithmetic_simd.mojo, extensor.mojo, test_gradient_clipping.mojo) |
| Agents spawned | 0 |

**Comment by mvillmow** [comment]
Created 3 follow-up issue(s): #5141, #5143, #5145
